### PR TITLE
Features/554 permutation

### PR DESCRIPTION
**Pull request by mtar**
## Description

This Pull Request introduces the functions random.permutation and random.randperm imitating the behaviour of the respective NumPy or PyTorch function.

Issue/s resolved: #554 

## Changes proposed:
- add `random.permutation` [NumPy documentation](https://numpy.org/devdocs/reference/random/generated/numpy.random.permutation.html)
- add `random.randperm` [PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.randperm.html)

## Type of change

New feature

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by mtar** [comment]
run tests

**Comment by mtar** [comment]
rerun tests

**Comment by mtar** [comment]
rerun tests

File: heat/core/random.py [hunk 1]
```diff
@@ -237,6 +239,83 @@ def __kundu_transform(values):
...
+    >>> ht.random.permutation(arr)
+    DNDarray([[3, 4, 5],
+              [6, 7, 8],
+              [0, 1, 2]], dtype=ht.int32, device=cpu:0, split=None)
+    """
+    if isinstance(x, int):
+        return randperm(x)
+
+    if not isinstance(x, dndarray.DNDarray):
+        raise TypeError("x must be integer or DNDarray")
```
Comment (coquelin77) [inline heat/core/random.py:279]: "x must be integer or DNDarray" -> "x must be an integer or DNDarray" or "x must be int or DNDarray"

File: heat/core/random.py [hunk 2]
```diff
@@ -237,6 +239,83 @@ def __kundu_transform(values):
...
+            requests.append(x.comm.Isend(x.lloc[i].clone(), proc_send, tag=tag))
+
+        for req in requests:
+            req.Wait()
+
+        data = torch.stack(buf)
+    else:
+        data = torch.empty_like(x._DNDarray__array)
+
+    return factories.array(data, dtype=x.dtype, is_split=x.split, comm=x.comm)
```
Comment (coquelin77) [inline heat/core/random.py:316]: device from `x` as well

File: heat/core/tests/test_random.py [hunk 3]
```diff
@@ -310,6 +348,25 @@ def test_randn(self):
         self.assertFalse(np.allclose(a, c))
         self.assertFalse(np.allclose(b, c))
 
+    def test_randperm(self):
+        state = torch.random.get_rng_state()
+
+        # results
```
Comment (coquelin77) [inline heat/core/tests/test_random.py:354]: can you test with float64 as well?

also should the return values should be `int64` correct? Can you test to make sure that they are what is expected?

**Review by coquelin77 — CHANGES_REQUESTED** [review]
These functions are very clean and easy to read. However, these solutions are still memory bound. This may be a problem in the future when a vector of length equal to `gshape[0]` cannot fit into memory. I am okay with adding this for now because I think this case is still a corner case, @Markus-Goetz what do you think?

File: heat/core/random.py (hunk 1 above)
Comment (mtar) [reply heat/core/random.py]: fixed

**Comment by Markus-Goetz** [comment]
Let's have a implementation first and the improve on it when necessary.

**Comment by mtar** [comment]
rerun tests

**Review by coquelin77 — APPROVED** [review]

**Pull request merged**